File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseMainFragment.java
package com.habitrpg.android.habitica.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.TabLayout;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.habitrpg.android.habitica.data.ApiClient;
import com.habitrpg.android.habitica.data.UserRepository;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.helpers.SoundManager;
import com.habitrpg.android.habitica.models.user.User;
import com.habitrpg.android.habitica.ui.activities.MainActivity;
import com.roughike.bottombar.BottomBar;

import javax.inject.Inject;

public abstract class BaseMainFragment extends BaseFragment {

    @Inject
    public ApiClient apiClient;
    @Inject
    protected
    UserRepository userRepository;
    @Nullable
    public MainActivity activity;
    @Nullable
    public TabLayout tabLayout;
    @Nullable
    private CollapsingToolbarLayout collapsingToolbar;
    @Nullable
    protected FrameLayout toolbarAccessoryContainer;
    @Nullable
    public BottomBar bottomNavigation;
    public ViewGroup floatingMenuWrapper;
    public boolean usesTabLayout;
    public boolean usesBottomNavigation = false;
    public int fragmentSidebarPosition;
    @Inject
    protected SoundManager soundManager;
    @Nullable
    protected User user;

    public void setUser(@Nullable User user) {
        this.user = user;
    }

    public void updateUserData(User user) {
        this.user = user;
    }

    public void setTabLayout(@Nullable TabLayout tabLayout) {
        this.tabLayout = tabLayout;
    }

    public void setBottomNavigation(@Nullable BottomBar bottomNavigation) {
        this.bottomNavigation = bottomNavigation;
    }

    public void setFloatingMenuWrapper(ViewGroup view) {
        this.floatingMenuWrapper = view;
    }

    public void setActivity(@Nullable MainActivity activity) {
        this.activity = activity;
    }

    @Override
    public void onViewStateRestored(Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (getActivity().getClass().equals(MainActivity.class)) {
            this.activity = (MainActivity) getActivity();
        }
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        if (savedInstanceState != null && savedInstanceState.containsKey("userId")) {
            String userId = savedInstanceState.getString("userId");
            if (userId != null && userRepository != null) {
                compositeSubscription.add(userRepository.getUser(userId).subscribe(habitRPGUser -> user = habitRPGUser, RxErrorHandler.handleEmptyError()));
            }
        }

        if (bottomNavigation != null) {
            if (this.usesBottomNavigation) {
                bottomNavigation.removeOnTabSelectListener();
                bottomNavigation.removeOnTabReselectListener();
                bottomNavigation.setVisibility(View.VISIBLE);
            } else {
                bottomNavigation.setVisibility(View.GONE);
            }
        }

        if (floatingMenuWrapper != null) {
            floatingMenuWrapper.removeAllViews();
        }

        setHasOptionsMenu(true);

        if (activity != null) {
            activity.setActiveFragment(this);
        }

        updateTabLayoutVisibility();

        return null;
    }

    private void updateTabLayoutVisibility() {
        if (tabLayout != null) {
            if (this.usesTabLayout) {
                tabLayout.removeAllTabs();
                tabLayout.setVisibility(View.VISIBLE);
                tabLayout.setTabMode(TabLayout.MODE_FIXED);
            } else {
                tabLayout.setVisibility(View.GONE);
            }
        }
    }

    @Override
    public void onDestroy() {
        if (userRepository != null) {
            userRepository.close();
        }
        super.onDestroy();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        if (user != null) {
            outState.putString("userId", user.getId());
        }

        super.onSaveInstanceState(outState);
    }

    public String customTitle() {
        return "";
    }

    public void hideToolbar() {
        activity.avatar_with_bars.setVisibility(View.GONE);
    }

    public void showToolbar() {
        activity.avatar_with_bars.setVisibility(View.VISIBLE);
    }

    public void disableToolbarScrolling() {
        if (collapsingToolbar != null) {
            AppBarLayout.LayoutParams params = (AppBarLayout.LayoutParams) collapsingToolbar.getLayoutParams();
            params.setScrollFlags(0);
        }
    }

    public  void enableToolbarScrolling() {
        if (collapsingToolbar != null) {
            AppBarLayout.LayoutParams params = (AppBarLayout.LayoutParams) collapsingToolbar.getLayoutParams();
            params.setScrollFlags(AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL | AppBarLayout.LayoutParams.SCROLL_FLAG_EXIT_UNTIL_COLLAPSED);
        }
    }

    public void setCollapsingToolbar(@Nullable CollapsingToolbarLayout collapsingToolbar) {
        this.collapsingToolbar = collapsingToolbar;
    }

    public void setToolbarAccessoryContainer(FrameLayout toolbarAccessoryContainer) {
        this.toolbarAccessoryContainer = toolbarAccessoryContainer;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseMainFragment.kt
package com.habitrpg.android.habitica.ui.fragments

import android.graphics.PorterDuff
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.WindowInsetsController
import androidx.core.os.bundleOf
import androidx.core.view.WindowCompat
import androidx.viewbinding.ViewBinding
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.tabs.TabLayout
import com.google.firebase.analytics.FirebaseAnalytics
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.data.UserRepository
import com.habitrpg.android.habitica.extensions.setNavigationBarDarkIcons
import com.habitrpg.android.habitica.helpers.SoundManager
import com.habitrpg.android.habitica.ui.activities.MainActivity
import com.habitrpg.android.habitica.ui.helpers.ToolbarColorHelper
import com.habitrpg.common.habitica.extensions.getThemeColor
import com.habitrpg.common.habitica.extensions.isUsingNightModeResources
import javax.inject.Inject

abstract class BaseMainFragment<VB : ViewBinding> : BaseFragment<VB>() {
    @Inject
    lateinit var userRepository: UserRepository

    @Inject
    lateinit var soundManager: SoundManager

    protected var showsBackButton: Boolean = false

    open val mainActivity get() = activity as? MainActivity
    val tabLayout get() = mainActivity?.binding?.content?.detailTabs
    val collapsingToolbar get() = mainActivity?.binding?.content?.toolbar
    val toolbarAccessoryContainer get() = mainActivity?.binding?.content?.toolbarAccessoryContainer
    val bottomNavigation get() = mainActivity?.binding?.content?.bottomNavigation
    var usesTabLayout: Boolean = false
        set(value) {
            field = value
            updateTabLayoutVisibility()
        }
    var hidesToolbar: Boolean = false
    var usesBottomNavigation = false

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        if (this.usesBottomNavigation) {
            bottomNavigation?.visibility = View.VISIBLE
        } else {
            bottomNavigation?.visibility = View.GONE
        }

        setHasOptionsMenu(true)

        tabLayout?.removeAllTabs()
        updateTabLayoutVisibility()
        updateToolbarInteractivity()

        if (hidesToolbar) {
            hideToolbar()
            disableToolbarScrolling()
        } else {
            showToolbar()
            enableToolbarScrolling()
        }
        context?.let {
            FirebaseAnalytics.getInstance(it).logEvent(
                "fragment_view",
                bundleOf(Pair("fragment", this::class.java.canonicalName))
            )
        }

        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onResume() {
        super.onResume()
        mainActivity?.showBackButton = showsBackButton
        mainActivity?.supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val window = activity?.window
        if (window != null) {
            val windowInsetsController = WindowCompat.getInsetsController(window, window.decorView)
            if (this.usesBottomNavigation) {
                windowInsetsController.isAppearanceLightNavigationBars = false
                view?.systemUiVisibility
                window.setNavigationBarDarkIcons(false)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    window.isNavigationBarContrastEnforced = false
                }
            } else {
                windowInsetsController.isAppearanceLightNavigationBars = requireActivity().isUsingNightModeResources()
                window.setNavigationBarDarkIcons(!requireActivity().isUsingNightModeResources())
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    window.isNavigationBarContrastEnforced = true
                }
            }
        }
    }

    @Deprecated("Use onCreateOptionsMenu(Menu, MenuInflater) instead")
    override fun onCreateOptionsMenu(
        menu: Menu,
        inflater: MenuInflater
    ) {
        super.onCreateOptionsMenu(menu, inflater)
        mainActivity?.toolbar?.let { ToolbarColorHelper.colorizeToolbar(it, mainActivity) }
        updateToolbarInteractivity()
    }

    var isTitleInteractive = false

    open fun updateToolbarInteractivity() {
        mainActivity?.updateToolbarInteractivity(isTitleInteractive)
    }

    private fun updateTabLayoutVisibility() {
        if (this.usesTabLayout) {
            tabLayout?.visibility = View.VISIBLE
            tabLayout?.tabMode = TabLayout.MODE_FIXED
        } else {
            tabLayout?.visibility = View.GONE
        }
    }

    override fun onDestroy() {
        userRepository.close()
        super.onDestroy()
    }

    private fun hideToolbar() {
        mainActivity?.binding?.content?.headerView?.visibility = View.GONE
    }

    private fun showToolbar() {
        mainActivity?.binding?.content?.headerView?.visibility = View.VISIBLE
    }

    private fun disableToolbarScrolling() {
        val params = collapsingToolbar?.layoutParams as? AppBarLayout.LayoutParams
        params?.scrollFlags = 0
    }

    private fun enableToolbarScrolling() {
        val params = collapsingToolbar?.layoutParams as? AppBarLayout.LayoutParams
        params?.scrollFlags =
            AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL or AppBarLayout.LayoutParams.SCROLL_FLAG_EXIT_UNTIL_COLLAPSED
    }

    protected fun tintMenuIcon(item: MenuItem?) {
        context?.getThemeColor(R.attr.headerTextColor)?.let {
            item?.icon?.setTint(it)
            item?.icon?.setTintMode(PorterDuff.Mode.MULTIPLY)
        }
    }
}
